[Blog] Introducing Extensions for OpenSearch: OpenSearch 2.9 (#1765)
opensearch-project/project-website: _posts/2023-07-10-introducing-extensions-for-opensearch.md

---
layout: post
title:  "Introducing Extensions for OpenSearch"
authors:
  - vemsarat
  - widdisd
  - kazabdu
  - dblock
  - dagney
  - minalsha
date: 2023-07-10
categories:
  - technical-posts
meta_keywords: OpenSearch, plugins, extensions

> meta_description: Introducing Extensions for OpenSearch, a new way to extend features of OpenSearch.

Reviewer: Updated meta and keywords: 

**Meta_keywords:**  OpenSearch Extensions, OpenSearch plugins, OpenSearch extensibility 
**Meta_description:** Discover how the experimental Extensions feature in OpenSearch 2.9 helps extend OpenSearch to independently scale workloads without impacting cluster availability.

---
layout: post
title:  "Introducing extensions for OpenSearch"
authors:
  - vemsarat
  - widdisd
  - kazabdu
  - dblock
  - dagney
  - minalsha
  - kolchfa
date: 2023-07-10
categories:
  - technical-posts
Meta_keywords: OpenSearch extensions, OpenSearch plugins, OpenSearch extensibility
Meta_description: Discover how the experimental extensions feature in OpenSearch 2.9 helps extend OpenSearch to independently scale workloads without impacting cluster availability.
---

> _Extensions_ is a new experimental feature in OpenSearch 2.9 that allows you to extend OpenSearch without impacting cluster availability and to independently scale workloads. In this blog post, we will introduce extensions and compare them to plugins. Using a 36-node cluster running a machine learning algorithm that performs high-cardinality anomaly detection, we'll demonstrate how we have achieved a cost reduction of 33% per data node, with performance matching that of a plugin and only the added cost of one extension node.

Reviewer 2: First sentence, how about: "that allows you to extend OpenSearch and independently scale workloads, without impacting cluster availability"

Reviewer 3: Sounds good, but no comma after "workloads" :-)

## The old way: Plugins

> OpenSearch is a fork of Elasticsearch 7.10, which offered the extension of features through plugins. Until now, plugins were the best way to extend features in OpenSearch, and we’ve written a [blog post](https://opensearch.org/blog/plugins-intro/) to help you understand how they worked. The OpenSearch Project currently maintains 17 plugins, and we have learned a lot developing and operating them at scale over the last 2 years. Over this time, we’ve encountered many bottlenecks in the existing plugin architecture.

Reviewer 2: how they work* not worked. They are still working.

Plugins are class-loaded into memory during OpenSearch bootstrap and therefore run within the OpenSearch process. This leads to the following three architectural problems:

1. Plugins require rigid version compatibility with OpenSearch. Every patch version change in OpenSearch core requires the plugin to be recompiled and rereleased.
2. Plugins are not isolated, so they can fatally impact the OpenSearch cluster and cannot scale independently. 
3. Having to manage the dependencies of all the components together makes it challenging to include 20+ plugins in a single product distribution.

## The new way: Extensions

> Unlike plugins, extensions run as a sidecar process or on a remote node, are isolated, and have cleaner, well-defined interfaces that interact with OpenSearch core. These interfaces are versioned (they follow semantic versioning) and are guaranteed to work across minor and patch versions, providing the future possibility of being compatible with multiple major versions of OpenSearch.

Reviewer 4: ```suggestion
Unlike plugins, extensions run as an independent process or on a remote node, are isolated, and have cleaner, well-defined interfaces that interact with OpenSearch core. These interfaces are versioned (they follow semantic versioning) and are guaranteed to work across minor and patch versions, providing the future possibility of being compatible with multiple major versions of OpenSearch.
```

The extensions architecture is illustrated in the following figure. For details, see [extension design documentation](https://opensearch-project.github.io/opensearch-sdk-java/DESIGN.html).

![Extensions for OpenSearch](/assets/media/blog-images/2023-07-10-introducing-extensions-for-opensearch/extensions.png){: .img-fluid}

## Experimental SDK launch

> The first step on our [extensibility roadmap](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) is to launch a developer SDK that makes extending OpenSearch easier than with plugins. Today we are launching a Java version, [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java), that works with OpenSearch 2.9 or above. To turn on the experimental extensions feature, set `opensearch.experimental.feature.extensions.enabled` to `true` in your `opensearch.yml` file.

Reviewer 2: "and above" rather than "or above:

Reviewer 3: Let's use "and later".

> The first step on our [extensibility roadmap](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) is to launch a developer SDK that makes the experience of extending OpenSearch easier than with plugins. Today we are launching a Java version, [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java), that works with OpenSearch 2.9 and later. To turn on the experimental extensions feature, set `opensearch.experimental.feature.extensions.enabled` to `true` in your `opensearch.yml` file.

Reviewer 4: ```suggestion
The first step on our [extensibility roadmap](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) is to launch a SDK that makes the experience of extending OpenSearch easier than with plugins. Today we are launching a Java version, [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java), that works with OpenSearch 2.9 and later. To turn on the experimental extensions feature, set `opensearch.experimental.feature.extensions.enabled` to `true` in your `opensearch.yml` file.
```

> The first step on our [extensibility roadmap](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) is to launch a developer SDK that makes the experience of extending OpenSearch easier than with plugins. Today we are launching a Java version, [opensearch-sdk-java](https://github.com/opensearch-project/opensearch-sdk-java), that works with OpenSearch 2.9 and later. To turn on the experimental extensions feature, set `opensearch.experimental.feature.extensions.enabled` to `true` in your `opensearch.yml` file.

Reviewer 4: We should include all the options to enable the experimental feature or just provide [this](https://github.com/opensearch-project/opensearch-sdk-java/blob/main/DEVELOPER_GUIDE.md#enable-the-extensions-feature-flag) link.

Reviewer 5: Done, thanks!

```json
GET /_extensions/_ad/detectors/results/_search
{
  "query": {
    "range": {
      "execution_start_time": {
        "gte": 1687456200000,
        "lte": 1687456260000
      }
    }
  },
  "track_total_hits": true
}
```

```json
{
    ...
    "hits": {
        "total": {
            "value": 1000000,
            "relation": "eq"
        },
        ...
    ...
}
```

The plugin limits these analysis rates to ensure stable cluster performance. After some fine-tuning (we increased simultaneous detector tasks from 10 to 96 and reduced the batch task piece interval from 5 to 3 seconds), the extension version of Anomaly Detection was able to successfully complete this analysis in 55 seconds. This isn’t quite an apples-to-apples comparison because the test in the previously mentioned blog post involved 1 million smaller models x 1 time slice each to generate 1 million results (we needed more time to implement support for real-time anomaly detection in the extension). However, we believe that we should be able to support 1M models on a single node or on a small number of nodes once extensions are capable of running in that environment.

> Because our extension ran on a remote node, we did observe an expected increase in API call latency. For example, the start/stop detector call went from 50--100ms to 200--300ms. These calls are marginal CRUD operations that are used to set up anomaly detection. The following figure provides a latency comparison between extensions and plugins. For more information, see [Performance testing for the Anomaly Detection extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/725#issuecomment-1597874382).

Reviewer 4: ```suggestion
Because our extension ran on a remote node, we did observe an expected increase in API call latency. For example, the start/stop detector call went from 50--100ms to 200--300ms. These calls are marginal CRUD operations that are used to set up Anomaly Detection. The following figure provides a latency comparison between extensions and plugins. For more information, see [Performance testing for the Anomaly Detection extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/725#issuecomment-1597874382).
```

## Improving performance

We proved that upgrading AD from plugin to extension could match existing performance for end users. But could we do better? We can, in the following three ways.  

### 1. Use a cheaper cluster

We found that by rewriting AD as an extension we could reduce the cost of our cluster and still meet the same performance bar. Because AD no longer runs on data nodes, we were able to replace all memory-optimized r5.2xlarge nodes with general purpose c5.2xlarge nodes for a 33% cost saving per data node for historical analysis, with the additional cost of one extension node.

We were also able to further decrease the number of data nodes in order to optimize searching the data and indexing the results. For example, since most data nodes in the original performance benchmark had insignificant usage other than indexing results, we tested a cluster with only 24 c5.2xlarge data nodes along with 1x r5.16xlarge extension node, a further 33% reduction. We believe that with further experimentation and fine tuning, you can find an even more optimal mix, depending on other resource requirements for the cluster.

### 2. Increase indexing throughput

> While we have not run this experiment, we believe you can further independently fine-tune the cluster by increasing sharding and achieve better throughput in indexing AD results.

Reviewer 6: "when indexing" sounds good to me.

> ### 3. Increase the rate of generating historical results

Reviewer 3: "Increase the rate of historical results generation"? "Increase the historical results generation rate"?

Reviewer 7: I'm fine with the second version.

AD’s historical HCAD restricts the rate of historical time period results, limits the use of AD to half the heap, and sets the maximum for detection at 10 simultaneous threads to protect the cluster from being maxed out. You can raise these limits by adding more CPUs or RAM, swapping space with SSDs independently from the rest of the cluster, or horizontally scaling extension nodes.

## Conclusion

> Extensions save cost and improve performance by right-sizing all resources to the task being performed. In our test, we did not need memory-optimized servers for the whole cluster to do historical AD, and therefore could achieve a 33% cost reduction of data nodes.

Reviewer 3: ```suggestion
Extensions save costs and improve performance by right-sizing all resources to the task being performed. In our test, we did not need memory-optimized servers for the whole cluster to perform historical AD and therefore could achieve a 33% cost reduction per data node.
```

> As a plugin, AD shares CPU and memory with OpenSearch I/O and must limit its own resource consumption in order to ensure the cluster’s stability. For example, in the AD Plugin HCAD tests, models were limited to use half of the heap, which itself was half of the node’s memory. In the AD extension scenario, it was possible to scale the heap up to consume most of the server’s memory and fully utilize 64 vCPUs dedicated only to AD tasks.

Reviewer 3: ```suggestion
As a plugin, AD shares CPU and memory with OpenSearch I/O and must limit its own resource consumption in order to ensure the cluster’s stability. For example, in the AD plugin HCAD tests, models were limited to using half of the heap, which itself was half of the node’s memory. In the AD extension scenario, it was possible to scale the heap to consume most of the server’s memory and fully utilize 64 vCPUs dedicated only to AD tasks.
```

Other existing plugins that could similarly benefit from this model include all plugins that have memory- or CPU-bound concerns, such as ML Commons, k-NN, and Reporting. This new technology also allows implementing new resource-heavy extensions, such as reranking, data transformation, big data processing, or video indexing.

## Next steps

> We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting in the [OpenSearch forum](https://forum.opensearch.org/) or opening issues in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!

Reviewer 3: ```suggestion
We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!
```

> We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!

Reviewer 2: We should add transparency here on the level of investment. We are excited to gain feedback, but need to be clear we don't have a timeline to make this generally available. 
**Recommendation** put this : 

We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Moving forward, in the short term we will shift our focus to using the learnings from the Extensibility project to enable extensibility in our OpenSearch Machine Learning offerings. The goal is to deliver a spec-based application framework where users will be able to create no-code ML-powered applications in OpenSearch, and ultimately deliver a simple one-click way for users to test, build and deploy ML-powered applications using OpenSearch. Additionally, before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!

Reviewer 3: Edited version: We are thrilled with the progress we have made with extensibility so far and with the early performance benchmarking results. Moving forward, in the short term we will shift our focus to using the learnings from the Extensibility project to enable extensibility in our OpenSearch Machine Learning (ML) offerings. The goal is to deliver a spec-based application framework where users will be able to create no-code ML-powered applications in OpenSearch and ultimately deliver a simple one-click way for users to test, build, and deploy ML-powered applications using OpenSearch. Additionally, before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [opensearch-sdk-java repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!
